halmeter: Replace deprecated functions

havardAasen opened the pull request:
Replaces deprecated functions. The new version works with GTK 2 and
GTK 3, the previous version was written for GTK 1.2

The most significant change is that TreeView has replaced CList.

If the change is approved I would appreciate if it went into the LinuxCNC master branch as well.

This should resolve part of the issue in linuxcnc/linuxcnc#864

Commit:
halscope: Replace deprecated gtk_menu functions

Commit:
halscope: Replace deprecated window functions

The new gtk_widget_set_size_request sets the minimum size.

Commit:
halscope: Replace File/Open and File/Save dialog window

This replaces the open/save dialog which is now called GtkFileChooser.

For the save dialog, it will now prompt if the user tries to overwrite
an existing file. I have also replaced the function "halscope_suffix"
with a built in file filter. This was also added for the "Save Data
File" dialog.

havardAasen commented:
I accidentally removed a feature, I'll see if I can add it back in again.

Commit:
halmeter: Port to GTK3

Replaces deprecated functions. halmeter will now only work with GTK3.

The most significant change is that TreeView has replaced CList.

havardAasen commented:
It was the "select the previously selected item" feature I had missed. I should be good now.

rene-dev commented:
Thanks for the PR, Im trying to fix the build system to build the parts which work with the provided gtk version.
have you looked at halmeter?

havardAasen commented:
I believe that the porting of halmeter is finished, and that it should build with both GTK2 and GTK3. Though these lines will break compiling with GTK2 if you compile it with deprecated symbols disabled.
```
#ifdef GTK2
    gtk_notebook_set_homogeneous_tabs(GTK_NOTEBOOK(notebk), TRUE);
#endif
```

rene-dev commented:
Sorry, Halscope not halmetet

havardAasen commented:
Aaah, ok

Since this PR didn't get merged and I didn't receive any feedback I stopped working on halscope. But I can pick it up again.

Except for the graph drawing which I haven't looked at, most is up and running. I can make some commits out of it it and add it to this branch.

rene-dev commented:
Ok, cool! Please do, like I said Im working on the build system. It can probably even go in master, I will take care of this.

rene-dev commented:
let me rephrase it: if you can make it work with gtk2 and 3 I will merge it into master. if only gtk3, I will merge into this branch.
gtk3 is preferred.

Commit:
halscope: Replace deprecated gtk_radio_button functions

Commit:
halscope: Replace threads, CList to TreeView

Port to GTK2, replace deprecated clist with TreeView

Commit:
halscope: Replace trigger source, CList to TreeView

Compatible with GTK2 and later, replace deprecated CList with TreeView

Commit:
halscope: Replace Select Channel Source, CList to TreeView

Replaces CList with a TreeView in the "Select Channel Source" dialog
window. Search functionality is built-in, and it now correctly finds the
previously selected item.

GTK2 and GTK3 compatible

havardAasen commented:
The force-push included some minor simplifications.
For the gtk3 part, it could be more gtk3, if we replaced all the boxes with a grid instead.

I have some commits for halscope coming, but it's slow going.

Commit:
halscope: Replace various signals and callbacks

Commit:
halscope: Replace various deprecated functions.

havardAasen commented:
e9a3dad  This is the change I am least satisfied with. Part since I broke the search function, which is built-in to TreeView, but can't be utilized since the change gets activated immediately and therefore closes the window. An easy fix, which also is identical to how the list in halmeter work, is to not accept a new item until you hit enter or double click on it.
It's also some code duplication with those to lists.

I haven't touched the drawing of the graphs.

rene-dev commented:
I tried your branch today, and could not get it to compile: https://pastebin.com/6pTNgznV

rene-dev commented:
still getting errors, are you compiling against gtk3? 

did you search and replace gtk+-2.0 to gtk+-3.0 in configure.ac?

```plain
In file included from hal/utils/scope.c:57:
hal/utils/scope_usr.h:61:5: error: unknown type name ‘GdkGC’
   61 |     GdkGC *disp_context;
      |     ^~~~~
hal/utils/scope_usr.h:66:5: error: unknown type name ‘GtkObject’
   66 |     GtkObject *zoom_adj;
      |     ^~~~~~~~~
hal/utils/scope_usr.h:68:5: error: unknown type name ‘GtkObject’
   68 |     GtkObject *pos_adj;
      |     ^~~~~~~~~
hal/utils/scope_usr.h:75:5: error: unknown type name ‘GtkObject’
   75 |     GtkObject *mult_adj;
      |     ^~~~~~~~~
hal/utils/scope_usr.h:115:5: error: unknown type name ‘GtkObject’
  115 |     GtkObject *scale_adj;
      |     ^~~~~~~~~
hal/utils/scope_usr.h:118:5: error: unknown type name ‘GtkObject’
  118 |     GtkObject *pos_adj;
      |     ^~~~~~~~~
hal/utils/scope_usr.h:147:5: error: unknown type name ‘GtkObject’
  147 |     GtkObject *level_adj;
      |     ^~~~~~~~~
hal/utils/scope_usr.h:150:5: error: unknown type name ‘GtkObject’
  150 |     GtkObject *pos_adj;
      |     ^~~~~~~~~
hal/utils/scope_usr.h:170:5: error: unknown type name ‘GdkDrawable’
  170 |     GdkDrawable *win;  /* the window */
      |     ^~~~~~~~~~~
hal/utils/scope_usr.h:171:5: error: unknown type name ‘GdkColormap’
  171 |     GdkColormap *map;  /* the colormap for the window */
      |     ^~~~~~~~~~~
hal/utils/scope_usr.h:178:5: error: unknown type name ‘GdkGC’
  178 |     GdkGC *context;  /* graphics context for drawing */
      |     ^~~~~
```

havardAasen commented:
I'm not finished at porting it to gtk3 it was a bit more work than I initially thought and I'm currently struggling with the coloring and drawing. Both needs to be changed to work with gtk3. Up until and including the last commit, it should still work with gtk2. 

As I said, the major issue is the drawing and coloring, there is a lot of deprecation warning but these should hopefully be trivial to fix.

rene-dev commented:
Ah, ok. Let me know when you are finished. I already fixed the build system. Thanks for your work!

Commit:
halscope: Change 'Select Channel Source' dialog

To select a new source, a row has to be double-clicked or selected and
pressed enter on the keyboard. This is a change, since previously it was
enough with a single click.
It is easier to search for the correct item, the functionality is also
identical to that in halmeter.

We stopped using the dialog_generic_t struct. Most of the functionality we
used is built in GTK now, so we're using that instead.

havardAasen commented:
This should now build with gtk 3 and  I guess it's in a functional state, but there is still some bugs that needs to be sorted. I hope most is explained in the commit messages

rene-dev commented:
with this patch I was able to compile:
```plain
diff --git a/src/hal/utils/meter.c b/src/hal/utils/meter.c
index c8249fb4d..a884db2b9 100644
--- a/src/hal/utils/meter.c
+++ b/src/hal/utils/meter.c
@@ -626,7 +626,7 @@ static void create_probe_window(probe_t * probe)
             GTK_POLICY_AUTOMATIC, GTK_POLICY_ALWAYS);
 
         /* create and set tabs in notebook */
-        label = gtk_label_new_with_mnemonic(tab_label_text[n]);
+        int label = gtk_label_new_with_mnemonic(tab_label_text[n]);
         gtk_widget_set_size_request(label, 70, -1);
         gtk_notebook_append_page(GTK_NOTEBOOK(probe->notebook), scrolled_window, label);
 
diff --git a/src/hal/utils/scope_vert.c b/src/hal/utils/scope_vert.c
index 0e9bc13b0..d2b4aa108 100644
--- a/src/hal/utils/scope_vert.c
+++ b/src/hal/utils/scope_vert.c
@@ -990,7 +990,7 @@ static gboolean dialog_select_source(int chan_num)
                 GTK_POLICY_AUTOMATIC, GTK_POLICY_ALWAYS);
 
         /* create and set tabs in notebook */
-        hbox = gtk_hbox_new(TRUE, 0);
+        int hbox = gtk_hbox_new(TRUE, 0);
         gtk_label_new_in_box(tab_label_text[n], hbox, TRUE, TRUE, 0);
         gtk_notebook_append_page(GTK_NOTEBOOK(vert->notebook), scrolled_window, hbox);

```
halmeter core dumps with sigsev on startup, halscope does the same when selecting the channel

havardAasen commented:
Sorry about that, I believe this commit should fix it. Messed up when I cherry-picked the commits.

rene-dev commented:
compiling and working!
only issue I can find is a scaling issue when selecting a signal in halscope.
![image](https://user-images.githubusercontent.com/707259/103160280-479bda80-47d3-11eb-944f-f3813be6049d.png)

havardAasen commented:
Added the remaining changes and squashed the last commits

havardAasen commented:
This should sort the worst issues. If you still want this PR, I would like to clean it up a bit before it's included. If it's not something more you want me to do? The only thing I feel I'm missing is filtering options for the file chooser. 
FWIW it seems that gtk2 is being shipped in Bullseye.

rene-dev commented:
yes, I am still interested, and would like to get it fully working and into master. I will test your latest changes.

Commit:
halscope: Fix deprecation warnings when compiling with GTK3

Commit:
halscope: Use inline css to set color for channel select buttons

Commit:
halscope: Use cairo to draw the horizontal trigger window

Dropping the gnome_canvas ifdef since this is deprecated

Commit:
halscope: Port the rest to GTK3. The code that didn't fit anywhere else

Commit:
halscope: Use cairo to draw the oscilloscope window

Change from GdkDrawable to cairo. This also uses GdkRGBA over GdkColor.

Commit:
halscope: Use accessor functions instead of direct access

rene-dev commented:
Thanks, I will test your changes. If they work well, I will merge them into master. Are you aware of problems? Is there anything left to do, or do you consider this finished?

havardAasen commented:
I manage to fix the issues I had. I also got it cleaned up a bit, though I'm open for suggestion. As far as the migration to gtk3 I think it's done.
Regarding other things to do, I believe there is certain things that could be simplified more, and there is the possibility to clean up the code, removing end of line whitespace, making it consistent with the code style etc. But nothing that's pressing.

rene-dev commented:
It seems to run great, I cannot find any issues. Thanks!
I created a PR on master. Please make any additional PRs there. https://github.com/LinuxCNC/linuxcnc/pull/1156
